src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala

```diff
@@ -72,9 +78,217 @@ trait WorkflowComponent {
+
+      def saveInputResolutions(values: Seq[SubmissionValidationValue], workflowId: UUID) = {
+        DBIO.seq(values.map { case (v) =>
+          v.value match {
+            case None => (submissionValidationQuery += marshalInputResolution(v, None, Some(workflowId)))
+            case Some(attr) =>
+              DBIO.sequence(attributeQuery.insertAttributeRecords(v.inputName, attr, workspaceContext.workspaceId)) flatMap { ids =>
+                (submissionValidationQuery += marshalInputResolution(v, Some(ids.head), Some(workflowId)))
+              }
+          }
+        }.toSeq: _*)
+      }
+
+      submissionQuery.loadSubmissionEntityId(workspaceContext.workspaceId, workflow.workflowEntity) flatMap { eId =>
+        (workflowQuery += marshalWorkflow(UUID.fromString(submissionId), workflow, eId)) andThen
+          saveInputResolutions(workflow.inputResolutions, UUID.fromString(workflow.workflowId))
+      }
+    }
+
+    def update(workspaceContext: SlickWorkspaceContext, submissionId: String, workflow: Workflow): ReadWriteAction[Workflow] = {
+      uniqueResult[WorkflowRecord](findWorkflowById(UUID.fromString(submissionId), UUID.fromString(workflow.workflowId))) flatMap {
+        case None => throw new RawlsException(s"workflow does not exist: ${workflow}")
+        case Some(rec) =>
+          delete(workspaceContext, submissionId, workflow.workflowId) andThen
+            save(workspaceContext, submissionId, workflow) andThen
+            DBIO.successful(workflow)
+      }
+    }
+
+    def loadWorkflow(submissionId: UUID, workflowId: UUID): ReadAction[Option[Workflow]] = {
+      uniqueResult[WorkflowRecord](findWorkflowById(submissionId, workflowId)) flatMap {
+        case None => DBIO.successful(None)
+        case Some(rec) =>
+          for {
+            entity <- loadWorkflowEntity(rec.workflowEntityId)
+            inputResolutions <- loadInputResolutions(rec.id)
+            messages <- loadWorkflowMessages(rec.id)
+          } yield(Option(unmarshalWorkflow(rec, entity, inputResolutions, messages)))
+      }
+    }
+
+    private def loadWorkflowEntity(entityId: Option[Long]): ReadAction[Option[AttributeEntityReference]] = {
+      entityId match {
+        case None => DBIO.successful(None)
+        case Some(id) => uniqueResult[EntityRecord](entityQuery.findEntityById(id)).flatMap { rec =>
+          DBIO.successful(unmarshalEntity(rec))
+        }
+      }
+    }
+
+    def loadWorkflowMessages(workflowId: UUID): ReadAction[Seq[AttributeString]] = {
+      findWorkflowMessagesById(workflowId).result.flatMap(recs => DBIO.sequence(recs.map { rec =>
+        DBIO.successful(unmarshalWorkflowMessage(rec))
+      }))
+    }
+
+    def loadInputResolutions(workflowId: UUID): ReadAction[Seq[SubmissionValidationValue]] = {
+      findInputResolutionsByWorkflowId(workflowId).result.flatMap(recs => DBIO.sequence(recs.map { rec =>
```

Comment:
this can be done in a single join instead of calling .result flatMap to DBIO.sequence